### ISSUE TITLE
Image global counter in evaluate_batch?

### ISSUE THREAD

**Issue by drozzy** [determined-ai/determined#3536]
What index/step/count should I use when plotting images to the tensorboard in validation `evaluate_batch` function?

Is there a way to access some global batch number during evaluate_batch?
Additionally, I would like to plot images only on the first batch of validation dataset. Is that not possible?

Thank you.

**Comment by rb-determined-ai** [comment]
Sorry @drozzy we didn't see this message a bit earlier.

I'm guessing you are using the PyTorchTrial API?  As far as I know that is the only `evaluate_batch()` we have, and it comes with a [batch_idx parameter](https://docs.determined.ai/latest/training-apis/api-pytorch-reference.html#determined.pytorch.PyTorchTrial.evaluate_batch).  Is that not what you are looking for?

**Comment by drozzy** [comment]
I looks like `batch_idx` resets to 0 every time evaluate_batch is run. Whereas I would like to plot labels on validation set throughout the training process.

**Comment by rb-determined-ai** [comment]
Ah.  Why not save the `batch_idx` you get in `train_step` as `self.last_train_step`?  You could then read it in `evaluate_batch()`.

**Comment by rb-determined-ai** [comment]
Sounds to me like you'll also want a `self.latest_image` flag, and something like this:

```python
class MyTrial(PyTorchTrial):
    def __init__(self, context):
        self.context = context
        self.latest_batch = 0
        self.latest_image = None
        ...

    def train_batch(self, batch, epoch_idx, batch_idx):
        self.latest_batch = batch_idx
        ...
        
    def evaluate_batch(self, batch, batch_idx):
        # if you want only one worker to do this in distributed training
        if self.context.distributed.rank == 0:
            # if you want to only do this once per validation
            if self.latest_image != self.latest_batch:
                # GENERATE/SAVE IMAGE HERE
                self.latest_image = self.latest_batch
```

**Comment by drozzy** [comment]
> Ah. Why not save the `batch_idx` you get in `train_step` as `self.latest_batch`? You could then read it in `evaluate_batch()`.

Thanks, this is what I ended up doing. Just thought maybe there was a more "official" way. 

Do I actually need to check for `rank=0`? What I did was check that `batch_idx` in evaluate_batch was equal to 0. 
Could there be a scenario where in distributed setting, both rank=0 and rank=1 workers get a batch_idx=0 in evaluate_batch?

**Comment by rb-determined-ai** [comment]
> Could there be a scenario where in distributed setting, both rank=0 and rank=1 workers get a batch_idx=0 in evaluate_batch?

I'm pretty sure this is the case every time, but maybe run a distributed training experiment and and print `batch_idx` every batch to see if I'm just crazy.

**Comment by rb-determined-ai** [comment]
closing due to inactivity
